File: 1. JPA basic/basic/src/main/java/study/jpa/basic/jpashop/domain/Item.java
package study.jpa.basic.jpashop.domain;

import java.util.ArrayList;
import java.util.List;

import javax.persistence.Column;
import javax.persistence.DiscriminatorColumn;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;
import javax.persistence.ManyToMany;
import javax.persistence.Table;

/**
 * @author Joonhyuck Hyoung
 */
@Entity(name = "jpashop_item")
@Table(name = "jpashop_item")
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn // default = DTYPE
public abstract class Item extends BaseEntity{
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    // @Column(name = "MEMBER_ID") 회사 룰대로 가면 된다.
    private Long no;

    private String name;

    private int price;

    /**
     * 원래는 hibernate가 orderDate -> order_date 로 자동변환해준다.
     * 지금은 안됨?
     * https://www.baeldung.com/hibernate-naming-strategy
     */
    @Column(name = "stock_quantity")
    private int stockQuantity;

    @ManyToMany(mappedBy = "items")
    private List<Category> categories = new ArrayList<>();

    public void setNo(Long no) {
        this.no = no;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setPrice(int price) {
        this.price = price;
    }

    public void setStockQuantity(int stockQuantity) {
        this.stockQuantity = stockQuantity;
    }

    public List<Category> getCategories() {
        return categories;
    }

    public void setCategories(List<Category> categories) {
        this.categories = categories;
    }

    public Long getNo() {
        return no;
    }

    public String getName() {
        return name;
    }

    public int getPrice() {
        return price;
    }

    public int getStockQuantity() {
        return stockQuantity;
    }
}

File: 2. JPA deep/jpashop/src/main/kotlin/jpabook/jpashop/domain/item/Item.kt
package jpabook.jpashop.domain.item

import jpabook.jpashop.domain.*
import jpabook.jpashop.exception.*
import javax.persistence.*

@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "dtype")
open class Item constructor(
    val name: String,

    val price: Int,

    var stockQuantity: Int,
) {
    @Id
    @GeneratedValue
    @Column(name = "item_id")
    val id: Long = 0L

    @ManyToMany(mappedBy = "items")
    val categories: MutableList<Category> = arrayListOf()

    //== 비즈니스 로직 ==//
    fun addStock(quantity: Int) {
        stockQuantity += quantity
    }

    fun removeStock(quantity: Int) {
        (stockQuantity - quantity).also { restStock ->
            if(restStock < 0) throw NotEnoughStockException("need more stock")
            stockQuantity = restStock
        }
    }
}
